lib/synapse/config_generator/haproxy.rb

```diff
@@ -893,12 +893,22 @@ def generate_config(watchers)
       new_config = generate_base_config
       shared_frontend_lines = generate_shared_frontend
 
+      @frontends_cache ||= {}
+      @backends_cache ||= {}
       watchers.each do |watcher|
         watcher_config = watcher.config_for_generator[name]
         @watcher_configs[watcher.name] ||= parse_watcher_config(watcher)
         next if watcher_config['disabled']
-        new_config << generate_frontend_stanza(watcher, @watcher_configs[watcher.name]['frontend'])
-        new_config << generate_backend_stanza(watcher, @watcher_configs[watcher.name]['backend'])
+
+        if watcher.updated || @frontends_cache[watcher.name].nil?
+          @frontends_cache[watcher.name] = generate_frontend_stanza(watcher, @watcher_configs[watcher.name]['frontend'])
+        end
+        if watcher.updated || @backends_cache[watcher.name].nil?
+          @backends_cache[watcher.name] = generate_backend_stanza(watcher, @watcher_configs[watcher.name]['backend'])
+        end
+        new_config << @frontends_cache[watcher.name] << @backends_cache[watcher.name]
+        watcher.updated = false if watcher.updated
```

Comment:
I'm not a big fan of config_generator mutating service_watchers, they really need to be decoupled as much as possible IMO. Right now they have read only access to backends and config_for_generator ... but this would be adding complexity which is only present in haproxy (and might interfere with nginx or file_output).

Perhaps we can set this in the main loop after update_config is called on all config_generators? Or we can just have the config generator be 100% responsible for deciding to update or not (e.g. if backends haven't changed).